lib/chezmoi/scripts.go

```diff
@@ -0,0 +1,93 @@
+package chezmoi
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"regexp"
+)
+
+var onceRegex = regexp.MustCompile(`(?m)^#\s*chezmoi:\s*once$`)
+
+// Script is a script supposed to run.
+type Script struct {
+	Name             string
+	SourcePath       string
+	once             bool
+	alreadyExecuted  bool
+	contents         []byte
+	contentsErr      error
+	evaluateContents func() ([]byte, error)
+}
+
+// Apply executes the script if necessary to reach target state.
+func (s *Script) Apply(destDir string, dryRun bool) error {
+	if s.once && s.alreadyExecuted {
+		return nil
+	}
+
+	s.Contents()
+	if s.contentsErr != nil {
+		return s.contentsErr
+	}
+
+	if err := s.parse(); err != nil {
+		return err
+	}
+	fmt.Printf("chezmoi: Running script %s\n", s.Name)
+	if dryRun {
+		fmt.Print(string(s.contents), "\n\n")
+		return nil
+	}
+	return s.execute(destDir)
+}
+
+func (s *Script) parse() error {
+	if s.contentsErr != nil {
+		return s.contentsErr
+	}
+	s.once = onceRegex.Match(s.contents)
+	return nil
+}
+
+func (s *Script) execute(destDir string) error {
+	f, err := ioutil.TempFile(os.TempDir(), s.Name)
+	if err != nil {
+		return err
+	}
+	// We can ignore the error, the file is in a temporary dir anyways.
+	defer os.Remove(f.Name())
+	if _, err := f.Write(s.contents); err != nil {
+		return err
+	}
+	if err := f.Chmod(os.FileMode(0700)); err != nil {
+		return err
+	}
+	c := exec.Command(f.Name())
+	c.Dir = destDir
+	c.Env = append(os.Environ(), "CHEZMOI_SCRIPT_DIR="+path.Dir(s.SourcePath))
+	c.Stdout = os.Stdout
+	c.Stderr = os.Stderr
+	c.Stdin = os.Stdin
```

Comment:
Please use `c.Stdout()` and friends here. This was only recently added, but facilitates testing.